python/GafferOSLTest/OSLShaderTest.py

```diff
@@ -857,5 +857,47 @@ def testDisabledShaderPassesThroughExternalValue( self ) :
 		self.assertEqual( len( network ), 1 )
 		self.assertEqual( network[0].parameters["a"].value, IECore.V3f( 12, 11, 10 ) )
 
+	def testDisabledShaderEvaluatesStateCorrectly( self ) :
+
+		redShader = self.compileShader( os.path.dirname( __file__ ) + "/shaders/red.osl" )
+
+		n2 = GafferOSL.OSLShader( "red1" )
+		n2.loadShader(redShader)
+
+		n3 = GafferOSL.OSLShader( "red2" )
+		n3.loadShader(redShader)
+
+		n1 = GafferOSL.OSLShader( "add" )
+		n1.loadShader( self.compileShader( os.path.dirname( __file__ ) + "/shaders/add.osl" ) )
+
+		n1['parameters']['a'].setInput(n2["out"]["out"])
+		n1['parameters']['b'].setInput(n3["out"]["out"])
+
+		sphere = GafferScene.Sphere()
+
+		shaderAssignment  = GafferScene.ShaderAssignment()
+		shaderAssignment["in"].setInput(sphere["out"])
+
+		pathFilter = GafferScene.PathFilter()
+		pathFilter["paths"].setValue( IECore.StringVectorData( [ "/sphere" ] ) )
+
+		shaderAssignment["filter"].setInput(pathFilter["out"])
+		shaderAssignment["shader"].setInput(n1["out"]["out"])
+
+		attributes = shaderAssignment['out'].attributes("/sphere")
+		shader = attributes["osl:shader"]
+
+		self.assertEqual( len( shader ), 3 )
+
+		# when we disable the ColorMultiply we should get the pass through parameters shader which for ColorMultiply, which is 'a' and connected to n2.
+		n1["enabled"].setValue(False)
+
+		attributes = shaderAssignment['out'].attributes("/sphere")
+		shader = attributes["osl:shader"]
+
+		# There should be a single OSL shader after disabling n1 but because the effectiveParameter
```

Comment:
OK, let me try again. If I substitute in some placeholders, the sentence says "There should be X, but because Y". I assumed that meant the sentence was incomplete, and was intending to say something like "There should be X, but because Y, _something_.".

Hopefully it made sense when I said we should test that it's n1 that gets passed through and not n2?